layers/meta-balena-jetson/recipes-bsp/uefi/edk2-firmware-tegra/0001-Switch-boot-chains.patch

```diff
@@ -0,0 +1,213 @@
++         }
++         return Status;
++       }
++     }
++  }
++
++  return EFI_SUCCESS;
++}
++
++
+ /**
+   This is the declaration of an EFI image entry point. This entry point is
+   the same for UEFI Applications, UEFI OS Loaders, and UEFI Drivers, including
+@@ -2630,6 +2713,9 @@ L4TLauncher (
+   UINTN                      UpgradeAvailable = 0;
+   UINTN                      BootCountValue = 0;
+   UINTN                      RootFsIndex = BALENA_ROOTFS_INDEX_A;
++  UINTN                      UpdatedBootSlot = 0;
++  BOOLEAN                    RunningUpdatedBootSlot = FALSE;
++  BOOLEAN                    BootLimitReached = FALSE;
+   Status = gBS->HandleProtocol (ImageHandle, &gEfiLoadedImageProtocolGuid, (VOID **)&LoadedImage);
+   if (EFI_ERROR (Status)) {
+     ErrorPrint (L"%a: Unable to locate loaded image: %r\r\n", __FUNCTION__, Status);
+@@ -2711,14 +2797,35 @@ L4TLauncher (
+           Print (L"%a: BootCountValue is %d\r\n", __FUNCTION__, BootCountValue);
+       }
+ 
++      Status = ProcessBootSlotFile(LoadedImage->DeviceHandle, BootParams.BootChain, &UpdatedBootSlot, &RunningUpdatedBootSlot);
++      if (EFI_ERROR (Status)) {
++        ErrorPrint(L"%a: Error while processing boot slot value\r\n", __FUNCTION__);
++      } else {
++        ErrorPrint(L"%a: Updated (target) boot slot: %u - Current running boot slot: %u\r\n", __FUNCTION__, UpdatedBootSlot, BootParams.BootChain);
++      }
++
+       /* If boot limit is reached while upgrade_available=1, switch partitions */
+       if (BALENA_OS_BC_LIM <= BootCountValue) {
+-         if (BALENA_ROOTFS_INDEX_B == RootFsIndex) {
+-            RootFsIndex = BALENA_ROOTFS_INDEX_A;
+-         } else {
+-            RootFsIndex = BALENA_ROOTFS_INDEX_B;
+-         }
+-         ErrorPrint(L"%a: Reached bootcount limit, will switch partitions... New index is %d\r\n", __FUNCTION__, RootFsIndex);
++        if (BALENA_ROOTFS_INDEX_B == RootFsIndex) {
++          RootFsIndex = BALENA_ROOTFS_INDEX_A;
++        } else {
++          RootFsIndex = BALENA_ROOTFS_INDEX_B;
++        }
++        BootLimitReached = TRUE;
++        ErrorPrint(L"%a: Reached bootcount limit, will switch partitions... New index is %d\r\n", __FUNCTION__, RootFsIndex);
++
++        /* Running the same boot slot as our target, which has been updated by the UEFI capsule */
++        if (RunningUpdatedBootSlot) {
++          ErrorPrint(L"%a: Running the updated boot slot %u and boot limit was reached, rebooting into fallback (pre-update) slot %u\r\n", __FUNCTION__, BootParams.BootChain, !BootParams.BootChain);
++          Status = SetNextBootChain(!BootParams.BootChain);
++          ResetCold();
++        } else {
++          ErrorPrint(L"%a: Running the fallback (pre-update) boot slot %u because boot limit was reached. Updated boot slot booted previously was %u\r\n", __FUNCTION__, BootParams.BootChain, UpdatedBootSlot);
++          /* TODO: Check if this allows for capsule to be applied during rollback-altboot */

```

Comment:
ValidateActiveBootChain() does indeed allow for the rollback capsule to be applied on the next boot.